docs/design/phases.md

```diff
@@ -0,0 +1,354 @@
+ - a CA for the front proxy (`front-proxy-ca.crt`) with its key (`front-proxy-ca.key`)
+ - a client cert for the front proxy client (`front-proxy-client.crt`) using `front-proxy-ca.crt` as the CA with its key (`front-proxy-client.key`)
+
+
+If a given certificate and private key pair both exist, the generation step will be skipped and those files will be validated and used for the prescribed use-case.
+This means the user can, for example, give his own CA, which then will be used for signing the rest of the certs.
+
+### Generate KubeConfig files on the master
+
+There should be:
+ - a KubeConfig file for kubeadm to use itself and the admin: `/etc/kubernetes/admin.conf`
+   - the "admin" here is defined as `kubeadm` itself and the actual person(s) that is administering the cluster and want to control the cluster
+     - with this file, the admin has full control over the cluster 
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:masters` organization
+ - a KubeConfig file for kubelet to use: `/etc/kubernetes/kubelet.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - be in the `system:nodes` organization
+     - have the name `system:node:<hostname>`
+ - a KubeConfig file for controller-manager: `/etc/kubernetes/controller-manager.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - have the name `system:kube-controller-manager`
+ - a KubeConfig file for scheduler: `/etc/kubernetes/scheduler.conf`
+   - inside this file, a client certificate is generated from the `ca.crt` and `ca.key`. The client cert should:
+     - be a client certificate (`x509.ExtKeyUsageClientAuth`)
+     - have the name `system:kube-scheduler`
+
+`ca.crt` is also embedded in the KubeConfig files.
+
+### Bootstrap the control plane
+
+#### etcd
+
+Determine if the user specified external etcd options. If not, etcd should:
+ - be spun up as a Static Pod
+ - listen on `localhost:2379` and use `HostNetwork=true`
+ - have `PodSpec.SecurityContext.SELinuxOptions.Type=spc_t` set because of https://github.com/kubernetes/kubeadm/issues/107
+ - be at a minimum version `3.0.14`
+ - make a volume out to the host on a given host path
+
+#### API Server
+
+The API Server needs to know this in particular:
+ - The subnet to use for services
+ - Where to find the etcd server
+ - The address and port to bind to; defaults to the IP Address of the default interface and port 6443 for secure communication
+ - Any extra flags and/or HostPath Volumes/VolumeMounts specified by the user
+
+Other flags that are set unconditionally:
```

Comment:
Is there a single mast.conf file we can create?  which may turn into a component.config